File: verdeciel/Quest.swift

//  Created by Devine Lu Linvega on 2015-09-21.
//  Copyright © 2015 XXIIVV. All rights reserved.

import UIKit
import QuartzCore
import SceneKit
import Foundation

class Quest
{
	var name:String!
	var location:Location!
	var predicate:() -> Bool
	var result:() -> Void
	var isCompleted:Bool = false
	
	init(name:String, location:Location! = nil, predicate:() -> Bool, result: () -> Void )
	{
		self.name = name
		self.location = location
		self.predicate = predicate
		self.result = result
	}
	
	func validate()
	{
		if predicate() {
			complete()
		}
		else{
			isCompleted = false
		}
	}
	
	func complete()
	{
		if isCompleted == true { return }
		print("+ QUEST    | Complete: \(name!)")
		isCompleted = true
		self.result()
	}
}

File: verreciel_js/scripts/objects/quest.js
//  Created by Devine Lu Linvega.
//  Copyright © 2017 XXIIVV. All rights reserved.

class Quest
{
  constructor(name, location, predicate, result)
  {
    // assertArgs(arguments, 4);
    this.isCompleted = false;
    this.isSkipped = false;

    this.name = name;
    this.location = location;
    this.predicate = predicate;
    this.result = result;
  }

  validate()
  {
    // assertArgs(arguments, 0);
    if (this.isSkipped == true)
    {
      this.isCompleted = true;
    }
    else if (this.predicate())
    {
      this.complete();
    }
    else
    {
      this.isCompleted = false;
    }
  }
  
  complete()
  {
    // assertArgs(arguments, 0);
    if (this.isCompleted == true)
    {
      return;
    }
    this.isCompleted = true;
    this.result();
  }
}
